nominal/sdk.py

```diff
@@ -0,0 +1,508 @@
+        """Add attachments that have already been uploaded to this run."""
+        request = scout_run_api.UpdateAttachmentsRequest(
+            attachments_to_add=list(attachment_rids), attachments_to_remove=[]
+        )
+        self._run_client.update_run_attachment(self._auth_header, request, self.rid)
+
+    def remove_attachments(self, attachment_rids: Iterable[str]) -> None:
+        """Remove attachments from this run.
+        Does not remove the attachments from Nominal.
+        """
+        request = scout_run_api.UpdateAttachmentsRequest(
+            attachments_to_add=[], attachments_to_remove=list(attachment_rids)
+        )
+        self._run_client.update_run_attachment(self._auth_header, request, self.rid)
+
+    def replace(
+        self,
+        *,
+        title: str | None = None,
+        description: str | None = None,
+        properties: Mapping[str, str] | None = None,
+        labels: Sequence[str] | None = None,
+    ) -> Self:
+        """Replace run metadata.
+        Returns the run with updated metadata.
+        Only the metadata passed in will be replaced, the rest will remain untouched.
+
+        Note: This replaces the metadata rather than appending it. To append to labels or properties, merge them before
+        calling this method. E.g.:
+
+            new_labels = ["new-label-a", "new-label-b"]
+            for old_label in run.labels:
+                new_labels.append(old_label)
+            run = run.replace(labels=new_labels)
+        """
+        request = scout_run_api.UpdateRunRequest(
+            description=description,
+            labels=None if labels is None else list(labels),
+            properties=None if properties is None else dict(properties),
+            title=title,
+        )
+        response = self._run_client.update_run(self._auth_header, request, self.rid)
+        return self.__class__._from_conjure(self._auth_header, self._run_client, response)
+
+    @classmethod
+    def _from_conjure(cls, auth_header: str, run_client: scout.RunService, run: scout_run_api.Run) -> Self:
+        return cls(
+            rid=run.rid,
+            title=run.title,
+            description=run.description,
+            properties=MappingProxyType(run.properties),
+            labels=tuple(run.labels),
+            start=_conjure_time_to_integral_nanoseconds(run.start_time),
+            end=(_conjure_time_to_integral_nanoseconds(run.end_time) if run.end_time else None),
+            _auth_header=auth_header,
+            _run_client=run_client,
+        )
+
+
+@dataclass(frozen=True)
+class Dataset:
+    rid: str
+    name: str
+    description: str | None
+    properties: Mapping[str, str]
+    labels: Sequence[str]
+    _auth_header: str = field(repr=False)
+    _catalog_client: scout_catalog.CatalogService = field(repr=False)
+
+    def poll_until_ingestion_completed(self, dataset_rid: str, interval: timedelta = timedelta(seconds=2)) -> None:
```

Comment:
Oops, fallout from moving this from a function to a method; this should just use `self.rid` rather than passing it in